File: src/io/H5MDReader.py
#  Copyright (C) 2016
#      Max Planck Institute for Polymer Research & JGU Mainz
#
#  This file is part of ESPResSo++.
#
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.


r"""
*********************************************
**H5MDReader** - IO Object
*********************************************

* `write()`
  write configuration to H5MD file. By default filename is "out.h5",
  coordinates are folded.

  Properties

* `filename`
  Name of trajectory file. By default trajectory file name is "out.h5"

usage:

writing down trajectory

>>> read_conf_h5md = espressopp.io.H5MDReader(filename='trajectory.h5')
>>> for i in range (200):
>>>   integrator.run(10)
>>>   dump_conf_h5md.write()

reading trajectory

>>> dump_conf_h5md = espressopp.io.H5MDReader(filename='trajectory.h5')
>>> ext_analyze = espressopp.integrator.ExtAnalyze(dump_conf_h5md, 10)
>>> integrator.addExtension(ext_analyze)
>>> integrator.run(2000)

Both examples will give the same result: 200 configurations in trajectory .h5 file.

setting up length scale

For example, the Lennard-Jones model for liquid argon with :math:`\sigma=0.34 [nm]`

>>> dump_conf_h5md = espressopp.io.H5MDReader(system, integrator, filename='trj.h5', iomode=1, data_to_store=['pid', 'mass', 'velocity'], unfolded=False, length_factor=0.34, length_unit='nm', append=True)

will produce trj.h5 with  in nanometers // Federico P. comment: what in nanometers? It's clear coordinate but please don't leave sentence hanging!

.. function:: espressopp.io.H5MDReader(filename)

        :param filename: (default: 'out.h5')

        :type filename: string

.. function:: espressopp.io.H5MDReader.read()

        :rtype:
"""

import os
from espressopp.esutil import cxxinit
from espressopp import pmi
import espressopp

#from espressopp.ParticleAccess import *
from _espressopp import io_H5MDReader

class H5MDReaderLocal(io_H5MDReader):

  def __init__(self, filename):
    cxxinit(self, io_H5MDReader, filename)

  def read(self):
    if pmi.workerIsActive():
      self.cxxclass.read(self)


if pmi.isController :
  class H5MDReader(ParticleAccess):
    __metaclass__ = pmi.Proxy
    pmiproxydefs = dict(
      cls =  'espressopp.io.H5MDReaderLocal',
      pmicall = [ 'read' ],
      pmiproperty = ['filename']
    )

File: src/io/H5MDReader.cpp
/*
  Copyright (C) 2016
      JGU Mainz

  This file is part of ESPResSo++.

  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/


// http://stackoverflow.com/questions/10056393/g-with-python-h-how-to-compile
// http://realmike.org/blog/2012/07/08/embedding-python-tutorial-part-1/
#include "H5MDReader.hpp"  // keep python.hpp on top
#include "storage/Storage.hpp"
//#include "System.hpp"
#include "storage/DomainDecomposition.hpp"
#include "bc/BC.hpp"

#include "iterator/CellListIterator.hpp"

#include "boost/mpi.hpp"
#include "boost/mpi/communicator.hpp"
#include "mpi.h"

#include <fstream>
#include <sstream>

//#ifdef HDF5_LAYER
    //#include "hdf5.h"
    //#include "hdf5_hl.h"
//#endif

//#ifdef H5MD_LAYER
//    #include "ch5md.h"
//#endif

using namespace espressopp;
using namespace std;
using namespace boost::python;


namespace espressopp {
  namespace io {



    // Python wrapping
    void H5MDReader::registerPython() {

      using namespace espressopp::python;

      class_<H5MDReader, boost::noncopyable >
      ("io_H5MDReader", init< std::string>())
//        .add_property("filename", &H5MDReader::getFilename,
//                                  &H5MDReader::setFilename)
//	    .add_property("author", &H5MDReader::getFilename,
//								  &H5MDReader::setFilename)
//	    .add_property("author_email", &H5MDReader::getFilename,
//								  &H5MDReader::setFilename)
//	    .add_property("creator", &H5MDReader::getFilename,
//								  &H5MDReader::setFilename)
//		.add_property("creator_version", &H5MDReader::getFilename,
//								  &H5MDReader::setFilename)
//	    .add_property("iomode", &H5MDReader::getIomode,
//								&H5MDReader::setIomode)
//		.add_property("data_to_store", &H5MDReader::getDataToStore,
//										&H5MDReader::set_init_table)
//
//        .add_property("unfolded", &H5MDReader::getUnfolded,
//                                  &H5MDReader::setUnfolded)
//        .add_property("length_factor", &H5MDReader::getLengthFactor,
//                                       &H5MDReader::setLengthFactor)
//        .add_property("length_unit", &H5MDReader::getLengthUnit,
//                                     &H5MDReader::setLengthUnit)
//        .add_property("append", &H5MDReader::getAppend,
//                                  &H5MDReader::setAppend)
//        .def("write", &H5MDReader::write)
      ;
    } // end registerPython
  } // end namespace io
} // end namespace espressopp

